semtypes/src/main/java/io/ballerina/semtype/Common.java

```diff
@@ -114,22 +115,45 @@
         return Conjunction.and(atom, next);
     }
 
-    public static List<SemType> shallowCopyTypes(SemType[] v) {
-        return Arrays.asList(v);
+    public static SemType[] shallowCopyTypes(SemType[] v) {
+        return Arrays.copyOf(v, v.length);
     }
 
     public static List<SemType> shallowCopyTypes(List<SemType> v) {
         return new ArrayList<>(v);
     }
 
-    public static String[] shallowCopyTypes(String[] v) {
+    public static String[] shallowCopyStrings(String[] v) {
         return Arrays.copyOf(v, v.length);
     }
 
     public static boolean notIsEmpty(TypeCheckContext tc, SubtypeData d) {
         return false;
     }
 
+    // Returns whether s1.codePoints < s2.codePoints
+    public static boolean codePointCompare(String s1, String s2) {
+        if (s1.equals(s2)) {
+            return false;
+        }
+        int len1 = s1.length();
+        int len2 = s2.length();
+        if (len1 < len2 && s2.substring(0, len1).equals(s1)) {
+            return true;
+        }
+
+        for (int index = 0; index < len1 && index < len2; ) {
```

Comment:
Shouldn't this be running over codepoint count rather than lenght? And also shall we compare codepoints instead of chars at line:146

Reply:
fixed